### DISCUSSION TITLE
vips dzsave has "Too many files open"

### DISCUSSION THREAD

**Discussion by marcosocamargo** [libvips/libvips#3817] [Help wanted]
I am using vips dzsave and I am getting this error:
rename: unable to rename file "./kitten.jpg-WK3HH2/kitten.jpg_files" as "./kitten.jpg_files", Too many open files
I increased the ulimit to 32768 and I still get this error.
How much do I have to up the ulimit to make it work?
Is this normal or is it a bug.
Thanks,
-Marcos

**Comment by jcupitt** [comment]
Hi @marcosocamargo,

It should work. I see:

```
$ ulimit -n 20
$ vips dzsave st-francis.jpg x --vips-progress
vips temp-2: 30000 x 26319 pixels, 32 threads, 30000 x 16 tiles, 640 lines in buffer
vips temp-2: done in 7.04s          
```

So it needed fewer than 20 open files for a 30,000 x 26,319 pixel JPEG. This is libvips 8.15.1 on ubuntu 23.10.

What libvips version, what platform, how large is the file you are processing, etc.

**Comment by marcosocamargo** [comment]
I think it is related to renaming or moving the file to _files:

It is a smalll PNG 

-rw-r--r-- 1 root root 55672 Jan 15 17:37 kitten.png

fileroot@linux-web-server:/home/info/mount-folder# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye


root@linux-web-server:/home/info/mount-folder# vips -v
vips-8.10.5-Thu Jan 21 18:55:16 UTC 2021

This is Debian:
root@linux-web-server:/home/info/mount-folder# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye

**Comment by marcosocamargo** [comment]
root@linux-web-server:/home/info/mount-folder# vips dzsave kitten.png x --vips-progress
vips temp-2: 384 x 251 pixels, 1 threads, 384 x 1 tiles, 128 lines in buffer
vips temp-2: done in 7.57s
rename: unable to rename file "./x-QJNQH2/x_files" as "./x_files", Too many open files
root@linux-web-server:/home/info/mount-folder#

**Comment by marcosocamargo** [comment]
The folder where the file is located was a monted folder with gcsfuser (google storage). When I move the file to just a regular directory, it works.

root@linux-web-server:/home/info# vips dzsave kitten.png x --vips-progress
vips temp-2: 384 x 251 pixels, 1 threads, 384 x 1 tiles, 128 lines in buffer
vips temp-2: done in 0.0326s

**Reply by jcupitt** [reply]
Writing 1,000s of files to a network drive is probably not a good idea. You could write as zip instead:

```
root@linux-web-server:/home/info# vips dzsave kitten.png x.zip --vips-progress
```

It should be a lot faster.

**Comment by marcosocamargo** [comment]
It is a small file --384 x 251 pixels. This is just a test. How do know it is thousands of files?

**Reply by jcupitt** [reply]
But it will be 1,000s of small files if you try a real image, and I doubt if it'll work well, you'll see terrible performance and huge network overheads. If you write as a zip it'll be a single file write with no rename step. You could write to a local directory (`/tmp/x`?), then zip and copy over the network, but `dzsave` can do this for you very efficiently, so just writing a zip is much simpler.

This specific error you're seeing here is weird and possibly a bug in your network. Maybe it doesn't support remote rename and is being forced to copy and rewrite? It's hard to say.

**Reply by jcupitt** [reply]
Are you planning to use the pyramids for a slippy map viewer? A good solution here is to keep the pyramids in the zip and to change the `fetchTile()` method in the javascript to use http range requests to pull JPEGs directly from the zip. You'll find a huge drop in network overhead and the amount of server storage you need.

**Reply by marcosocamargo** [reply]
I am using the pyramids to display images of documents very much like what FamilySearch. When a user clicks on a link I open the image. What is slippy map viewer? I tried with zip and it works. I also talked to Google Storage and they are looking into it. I will try your suggestions. I am using openseadragon as the viewer. Can I pull the tile from the zip file using seadragon? fetchTile is part of vips? Thank you very much!

**Comment by marcosocamargo** [comment]
I have this error when opening image with seadragon ‘Access-Control-Allow-Origin’ is missing. Can Seadragon provide this in the header?

**Reply by jcupitt** [reply]
Sorry, no idea. Isn't that a problem with server config? I'm a bit out of my area here.

**Reply by marcosocamargo** [reply]
Thanks for all the help. Could you please point me to examples on how to use fetchTile()?

**Reply by jcupitt** [reply]
Oh it was just an example, you'll need to look into the openseadragon docs to see how they do it.

**Reply by marcosocamargo** [reply]
The image was rotated to the desired angle.  So I ran vips dzsave without any rotation and the image is back to its original rotation when the photo was taken. It seems that it doesn't recognize the rotation that was done. How do I fix that?

**Reply by jcupitt** [reply]
It's hard to say without seeing the image, but I would guess you did not actually rotate it, just set a image metadata item indicating the correct orientation. `dzsave` just tiles the image as-is and does not attempt to apply any metadata transforms.

In python you could use `autorot()` to apply rotation metadata transforms, eg.:

```
image.autorot().dzsave("x")
```

Though it will generally be much slower.

**Comment by marcosocamargo** [comment]
Is there any way I can find out what the metadata is before I tile it and then I can do the rotation using vips dzsave?
Thanks!

**Reply by jcupitt** [reply]
You can also do eg.:

```
vips autorot some-file-name output-name.dz[container=zip]
```

Or you can apply a rotation using a program that really will rotate the file and not just set a metadata tag.

**Comment by marcosocamargo** [comment]
How can I add quality = 50% and tile-size 510? I tried this command but it doesn't work
vips autorot IMG_238_0002.JPG y.dz[container=zip] --tile-size 510 --suffix .jpg[Q=50]

**Reply by jcupitt** [reply]
Try:

```
vips autorot IMG_238_0002.JPG y.dz[container=zip,tile_size=510,suffix=.jpg[Q=50]]
```

**Comment by marcosocamargo** [comment]
Basically I would like to autorotate using the EXIF tag rotate, zip file output, quality of jpeg 50% and tile size 510. How do I tell the tool to do that? Can I do that all in one command?
Thanks!

**Comment by marcosocamargo** [comment]
It worked. I really appreciate your help!

**Comment by marcosocamargo** [comment]
How do I prevent the tool from generating very small tiles of the document? These views are so small that they aren't usable. For example: https://storage.googleapis.com/recpres/y/y_files/4/0_0.jpg
If you look at the center of the screen, you will see a small thumbnail of the document.
Thanks again!

**Reply by jcupitt** [reply]
The deepzoom standard says you need levels all the way down to 1x1 pixels. You can use `depth=onetile` to make it stop when a level contains only one tile.

**Reply by marcosocamargo** [reply]
Thank you very much again!

**Reply by marcosocamargo** [reply]
I tried with one tile and I get this black box:
vips autorot /home/info/mount-folder/IMG_238_0002.JPG y.dz[container=zip,tile_size=510,suffix=.jpg[Q=50],depth=onetile]
Is my command correct?
Thanks!

**Reply by marcosocamargo** [reply]
![Capture](https://github.com/libvips/libvips/assets/156704677/82a24fc0-11ad-4a46-8414-da0fde61e316)

**Reply by jcupitt** [reply]
You need to post the source image somewhere or I won't be able to reproduce the problem.

**Reply by jcupitt** [reply]
Ah it's because you're not writing a deepzoom-compatible pyramid, so openseadragon can't display it. You must have all the pyramid layers, even the tiny ones.

If I run:

```shell
vips autorot page.jpg y.dz[suffix=.jpg[Q=50],tile_size=510]
```

It seems to work fine:

http://www.rollthepotato.net/~john/page/

**Comment by marcosocamargo** [comment]
![IMG_238_0002](https://github.com/libvips/libvips/assets/156704677/99a76c6a-90ac-4dd6-bf10-c57c40e6b6aa)

**Reply by marcosocamargo** [reply]
https://storage.googleapis.com/recpres/IMG_238_0002.JPG

**Reply by marcosocamargo** [reply]
Or you can download it from here:
https://drive.google.com/file/d/1AhJvCOZCEBTdzUZ_sVuEhk0BlpaX2wvj/view?usp=sharing

**Reply by marcosocamargo** [reply]
I noticed when with the zip option, I get the black rectangle. If I execute it without the zip option, then it works.

**Reply by jcupitt** [reply]
The zip option is turned on automatically for `.dz` output, you don't need to specify it.

**Comment by marcosocamargo** [comment]
Thanks again! Do you unzip the files or use it as is? Also, is there a reason why you omitted the tile size of 510?

**Reply by marcosocamargo** [reply]
When I don't turn on zip option, I don't get a zip option as you said
-Marcos

**Reply by jcupitt** [reply]
I unzipped for that test. You should get a zip file, it just doesn't have the `.zip` suffix:

```
$ vips copy ../k2.jpg x.dz
$ ls -l
total 460
-rw-r--r-- 1 john john 470065 Jan 20 16:39 x.dz
$ unzip x.dz
Archive:  x.dz
 extracting: x_files/11/0_0.jpeg     
 extracting: x_files/11/1_0.jpeg     
...
```

**Comment by marcosocamargo** [comment]
I issued this command:
vips autorot IMG_238_0002.JPG y.dz[suffix=.jpeg[Q=50],tile_size=510]
And I got these files which are not in zip format:

info@linux-web-server [~]# ls -l y*
-rw-r--r--  1 info info  206 Jan 20 16:58 y.dzi

y_files:
total 56
drwxr-xr-x 2 info info 4096 Jan 20 16:58 0
drwxr-xr-x 2 info info 4096 Jan 20 16:58 1
drwxr-xr-x 2 info info 4096 Jan 20 16:58 10
drwxr-xr-x 2 info info 4096 Jan 20 16:58 11
drwxr-xr-x 2 info info 4096 Jan 20 16:58 12
drwxr-xr-x 2 info info 4096 Jan 20 16:58 13
drwxr-xr-x 2 info info 4096 Jan 20 16:58 2
drwxr-xr-x 2 info info 4096 Jan 20 16:58 3
drwxr-xr-x 2 info info 4096 Jan 20 16:58 4
drwxr-xr-x 2 info info 4096 Jan 20 16:58 5
drwxr-xr-x 2 info info 4096 Jan 20 16:58 6
drwxr-xr-x 2 info info 4096 Jan 20 16:58 7
drwxr-xr-x 2 info info 4096 Jan 20 16:58 8
drwxr-xr-x 2 info info 4096 Jan 20 16:58 9

**Reply by jcupitt** [reply]
Ah you're on 8.10, I remember. It's changed a bit since then.

**Comment by marcosocamargo** [comment]
How do I install the latest version?
Thanks,
-Marcos

**Reply by jcupitt** [reply]
You don't need to, 8.10 should work fine, you just need to use the `container=zip` option.

I ran:

```
$ vips --version
vips-8.10.7-Sat Jan 20 20:12:52 UTC 2024
$ vips autorot ../page.jpg y.dz[container=zip,tile_size=510,suffix=.jpg[Q=50]]
$
```

Seems to work (after unzipping):

http://www.rollthepotato.net/~john/page/

**Comment by marcosocamargo** [comment]
If I want a tile size of 512 and an overlap of 1 pixel, is the overlap
included in the tile size?
For 512, the tile size is 510 and overlap 1, or should it be tile size 512
and overlap 1?
How does it work and what is the proper and common way of doing this?
Thanks!

**Reply by jcupitt** [reply]
`dzsave` used the deepzoom meanings for overlap and tilesize. 

If you want your tile images to be 512x512 pixels, set `tile-size` to 510 and `overlap` to 1:

```
$ vips dzsave ../k2.jpg x --tile-size 510 --overlap 1
$ vipsheader x_files/11/1_1.jpeg 
x_files/11/1_1.jpeg: 512x512 uchar, 3 bands, srgb, jpegload
```

Now the right column of tile 1x1 sticks 1 pixel into tile 2x1, and the left column of 2x1 sticks one pixel into 1x1. So there are two columns of pixels in the overlap.

Edge tiles have no overlap, so are one pixel smaller:

```
$ vipsheader x_files/11/0_1.jpeg 
x_files/11/0_1.jpeg: 511x512 uchar, 3 bands, srgb, jpegload
```

It's confusing to explain, but very simple if you draw a picture.

**Comment by marcosocamargo** [comment]
Hi John,
I appreciate  your help. What is the command that I need to issue in
PowerShell to get the same result as the one below? I am using version
8.15.1
I need to generate deepzoom files with a particular tile size and quality.
Also I need to auto rotate
.\vips dzsave autorot "C:\Users\James Camargo\Downloads\IMG_0120.jpg"
"C:\Users\James
Camargo\Downloads\IMG_0120.dz[tile_size=510,suffix=.jpeg[Q=50]]"
-Marcos Camargo

**Comment by marcosocamargo** [comment]
This seems to work but how do I add the autorotate to the command below?
.\vips dzsave "C:\Users\James Camargo\Downloads\IMG_0120.jpg"
 "C:\Users\James Camargo\Downloads\IMG_0120" --suffix .jpg[Q=50]
--tile-size 510
Thanks,
-Marcos

**Comment by marcosocamargo** [comment]
Also, I would like to remove any EXIF data from the resulting image.
Thanks.
-Marcos

**Comment by jcupitt** [comment]
Something like:

```
.\vips.exe dzsave "C:\Users\James Camargo\Downloads\IMG_0120.jpg[autorot]" "C:\Users\James Camargo\Downloads\IMG_0120" --suffix .jpg[Q=50] --tile-size 510
```

**Comment by marcosocamargo** [comment]
Thanks again!
The .dzi gets created but not all the file inside the directory
IMG_0120_files.
-Marcos

**Comment by marcosocamargo** [comment]
I am sorry. They were created.
-Marcos

**Comment by marcosocamargo** [comment]
Thanks again for your help. How do I make sure that every exif data is
removed from the resulting files?
-Marcos Camargo

**Reply by jcupitt** [reply]
`dzsave` always strips tiles of all optional metadata. You need to set eg. `keep=all` to keep it.

**Comment by marcosocamargo** [comment]
I tried the command it seems that it doesn't autorotate.
This is the command:
vips.exe dzsave "G:\Records Preservation\Brasil\São Paulo\São Paulo\Arquivo
Metropolitano Dom Duarte Leopoldo e Silva\Processos Gerais Antigos\Caixa
004\IMG_004_0709.JPG[autorot]" "E:\Brasil\São Paulo\São Paulo\Arquivo
Metropolitano Dom Duarte Leopoldo e Silva\Processos Gerais Antigos\Caixa
004\IMG_004_0709" --suffix .jpg[Q=50] --tile-size 510 --overlap 1

Thanks,
-Marcos

**Comment by marcosocamargo** [comment]
It works if you use autorotate instead of autorot
Thanks,
-Marcos

**Comment by marcosocamargo** [comment]
I am using openseadragon which doesn't load anything smaller than a tile. I used the onetile version of libvips which generates the necessary tiles for OSD but the numbering is incorrect. For example, if an image requires 13 levels down to 1 pixel, then with onetile option there are 5 levels from numbers 0-4 but OSD expects from 9-13. Is it possible to name the levels starting with highest levels 13 down to 9 instead of 0 to 4? Thanks!

**Reply by jcupitt** [reply]
Hi @marcosocamargo, if you add the `onetile` option, you get an invalid deepzoom image. The option is there for other layouts, like google maps. It's just an implementation detail that openseadragon won't bother loading smaller tiles.

I would write a regular deepzoom image (ie. no `onetile` option), then delete the smaller levels, if you're sure you won't need them.

**Comment by marcosocamargo** [comment]
I tested by generating the pyramid with the option onetile which generates
everything down to one tile. Then I renamed the directories in the _files
directory to be the name as if they were generated with onepixel option.
This works. Is there a way to add an option to onepixel to stop at one tile
level? That would save a lot of image generation especially for those that
use OSD? How do I determine which one to delete since every image is
different? How do you determine how many levels you are going to have?
Thanks for your help.
-Marcos

**Reply by jcupitt** [reply]
I don't think dzsave should support non-standard DZI images, and in any case the savings are tiny. For speed, dzsave generates several 1,000 tiles a second, so saving 7 tiles will not be measurably faster. The space saving will be only a few kb.

You can work out which directories OSD won't read by searching up from level 0 for the first level containing more than one tile, then backing off one.

For example:

```
$ vips dzsave nina.jpg x
$ ls x_files/*
vips-properties.xml

0:
0_0.jpeg

1:
0_0.jpeg

2:
0_0.jpeg

3:
0_0.jpeg

4:
0_0.jpeg

5:
0_0.jpeg

6:
0_0.jpeg

7:
0_0.jpeg

8:
0_0.jpeg

9:
0_0.jpeg  1_0.jpeg
```

So you can delete 0-7. It'll almost always be 0-7 for 256x256 tiles (ie. 2 ^ (8 - 1)), though maybe there are some strange edge cases where it'll be a different value.

**Comment by marcosocamargo** [comment]
Thanks for the input. Since I will be processing 1 million images it is
important not to generate anything that I don't need.
So I decided to write a script that generates down to 1 tile and then
renames the files based on going down to 1 pixel.
Just want to confirm that's how you calculate the maximum number of levels:
ceiling( ln (Max(height, width))/log 2
Thanks,
-Marcos

**Reply by jcupitt** [reply]
It shrinks the base level by *2, rounding up each time, then stops once the image fits within one tile. So yes I think that's the expression you quoted.

**Comment by marcosocamargo** [comment]
Thanks for the help again. I issued this command with has UTF8 characters:
vips.exe dzsave "E:\Brasil\São Paulo\Jundiaí\Nossa Senhora do
Desterro\Óbitos 1744-1787\IMG_007_0157_IRF.JPG" "E:\Brasil\São
Paulo\Jundiaí\Nossa Senhora do Desterro\Óbitos 1744-1787\IMG_007_0157"
--suffix .JPG[Q=80] --tile-size 512 --overlap 1 --depth onetile

And I get this error because the diacritics get messed (PowerShell)
VipsForeignLoad: file "E:\Brasil\SA£o Paulo\JundiaA-\Nossa Senhora do
Desterro\A"bitos 1744-1787\IMG_007_0157_IRF.JPG" does not exist

What kind of encoding does libvips accept? That might be the problem.

-Marcos

**Reply by jcupitt** [reply]
libvips is utf8, so I'd guess the problem is in powershell somewhere.

**Reply by jcupitt** [reply]
I tried on win10 and I see:

```
jcupi@DESKTOP-HGI6HBR MINGW64 /f/Pictures
$ ../vips-dev-8.15/bin/vips.exe text x.png 😁 --dpi 300 --rgba
```

To make:

![x](https://github.com/libvips/libvips/assets/580843/525e4b49-260e-4a04-8ec9-1f00f79e0602)


Filenames are different -- windows usually uses latin1 for filename encodings, though not always. libvips will try to convert windows filenames to utf8 automatically, but I think it may sometimes fail.

You could try the latest libvips, it might be better.

**Reply by jcupitt** [reply]
Or linux, of course, which is just utf8 everywhere.

**Comment by marcosocamargo** [comment]
I installed 8.15.2 and it is working now. Thanks for the suggestion.
Is there a thumbnail generator for VIPS?
-Marcos

**Comment by marcosocamargo** [comment]
I am reading about thumbnails here
https://www.libvips.org/API/current/Using-vipsthumbnail.html"

If the image has an EXIF tag to rotate it, will the thumbnail rotate it
too? This is my desired outcome, that is, that it rotates the thumbnail
according to the EXIF tag.
Thanks,
-Marcos

**Reply by jcupitt** [reply]
Yes, it will spin thumbnails upright.

**Comment by marcosocamargo** [comment]
vipsthumbnails doesn't seem to support UTF8. I have the latest version.
Thanks,
-Marcos

**Comment by marcosocamargo** [comment]
It seems that the input file name for both dzsave and thumbnails doesn't
take just a UTF8 defined in a variable but accepts it if I put it through
Get-Item in PowerShell.
Should it take both? I am using Irfanview in the same script and it takes
both ways.
Thanks,
-Marcos

**Comment by marcosocamargo** [comment]
vipsthumbnails doesn't work at with whether passing a variable with
diacritics or putting the variable through Get-Item.
